kotlin-analysis/src/main/kotlin/org/jetbrains/dokka/analysis/KotlinAnalysis.kt

```diff
@@ -34,20 +62,39 @@ class DokkaAnalysisConfiguration(
 )
 fun KotlinAnalysis(context: DokkaContext): KotlinAnalysis = KotlinAnalysis(context.configuration.sourceSets, context.logger)
 
-interface KotlinAnalysis : SourceSetDependent<EnvironmentAndFacade> {
-    override fun get(key: DokkaSourceSet): EnvironmentAndFacade
-    operator fun get(sourceSetID: DokkaSourceSetID): EnvironmentAndFacade
+interface KotlinAnalysis : SourceSetDependent<AnalysisContext>, Closeable {
+    override fun get(key: DokkaSourceSet): AnalysisContext
+    operator fun get(sourceSetID: DokkaSourceSetID): AnalysisContext
 }
 
-internal class KotlinAnalysisImpl(
-    private val environments: SourceSetDependent<EnvironmentAndFacade>
-) : KotlinAnalysis, SourceSetDependent<EnvironmentAndFacade> by environments {
+internal class ProjectKotlinAnalysis(
+    environments: SourceSetDependent<AnalysisContext>
+) : KotlinAnalysisImpl(environments)
+
+internal class SamplesKotlinAnalysis(
+    sampleEnvironments: SourceSetDependent<AnalysisContext>,
+    projectKotlinAnalysis: KotlinAnalysis
+) : KotlinAnalysisImpl(sampleEnvironments, projectKotlinAnalysis)
 
-    override fun get(key: DokkaSourceSet): EnvironmentAndFacade {
-        return environments[key] ?: throw IllegalStateException("Missing EnvironmentAndFacade for sourceSet $key")
+/**
+ * It does not close [defaultKotlinAnalysis].
+ */
+internal open class KotlinAnalysisImpl(
+    private val environments: SourceSetDependent<AnalysisContext>,
+    private val defaultKotlinAnalysis: KotlinAnalysis? = null
```

Comment:
Sorry, I didn't go into too much detail with my suggestion in the previous comment -- unfortunately, it's is still relevant.

The reason I suggested to separate the implementation into `ProjectKotlinAnalysis` and `SamplesKotlinAnalysis` is to avoid having an optional nullable field with a confusing name (`defaultKotlinAnalysis`).

It's unclear what `defaultKotlinAnalysis` is - it's default in relation to what? How is it limited to being "default"? As far as I see, it's just some kotlin analysis, can be anything.

```kotlin
internal class ProjectKotlinAnalysis(
    private val environments: SourceSetDependent<AnalysisContext>
)

internal class SamplesKotlinAnalysis(
    private val sampleEnvironments: SourceSetDependent<AnalysisContext>,
    private val projectKotlinAnalysis: KotlinAnalysis
)
```

^ Here you don't have that problem. `ProjectAnalysis` only has it's own environments. `SamplesAnalysis` has project's analysis and environments for samples, which is logical.

Additionally, if more optional/nullable fields get added, it'll be even more difficult to reason about (i.e in which case is field 1 null and field 2 not null or vise versa), which can be avoided early.

So my suggestion with separating the implementation  completely (without the base class) still stands, even if there's a little bit of "copy-paste".

Is there anything specific you don't like about it?

Reply:
Probably `defaultKotlinAnalysis`  is not a good name here. I will rename it to `fallbackKotlinAnalysis `.



>Additionally, if more optional/nullable fields get added, it'll be even more difficult to reason about (i.e in which case is field 1 null and field 2 not null or vise versa), which can be avoided early.

What do you mean? Why should we avoid nullable fields in Kotlin? As for me, it is ok to have no  `fallbackKotlinAnalysis here.`  

> Is there anything specific you don't like about it?

The current implementation is more general and does not have  "copy-paste".

Reply:
Discussed it separately and agreed on another solution, marking as resolved